metals/src/main/scala/scala/meta/internal/metals/ServerLivenessMonitor.scala

```diff
@@ -33,19 +38,30 @@ class RequestMonitor {
             outgoingMessage()
           case _: ResponseMessage => incomingMessage()
           case _: NotificationMessage => incomingMessage()
+          case _: RequestMessage => anyOutgoing()
           case _ =>
         }
         consumer.consume(message)
       }
 
     }
 
-  private def outgoingMessage() = lastOutgoing_ = now
-  private def incomingMessage(): Unit = lastIncoming_ = now
-  private def now = Some(System.currentTimeMillis())
+  private def outgoingMessage() = {
+    anyOutgoing()
+    lastOutgoing_ = Some(now)
+  }
+  private def anyOutgoing() = outgoing.getAndUpdate(_.appended(now))
```

Comment:
Could we instead of having the queue, check if Metals was idle at this point and set `lastIdle` time, later we would use the max(outgoingMessage, lastIdle) for checking server responsivenes?

Reply:
Yes, it is a far more reasonable solution.